In [ ]:
!pip install -q pyomo
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
#Creating a model
model = ConcreteModel()

#No. of decision variables
N = 14

#Objective function coefficient using numpy array
obj_coef_c = np.array([-4.2, 37.75, 3.5, 10.05, -8.25, 40.85, 0.05, -11.12, 0.65, 20.725, -102.65, 0.6, -9.45, -25.625])

#Constraints coeffiecient using numpy array
constr1_coef_p = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
constr2_coef_p = np.array([2.8, -0.6725, 3.6, 0, -2.02, 1.01, -4.305, 0, 1, -2.09, 0, -2.124, -3.402, 0])

#Declaring RHS of the constraints as variables
constr1_rhs_b1 = 343.48
constr2_rhs_b2 = 66.2

#Declaring lower and upper bounds of the constraints
lower_bound = np.array([2, -3, -np.inf, 1, 7, 3, -9, 0, 1, -25.245, -np.inf, -6, -30.5, -10.8])
upper_bound = np.array([+np.inf, 6, 2.5, 3.4, +np.inf, 6, 4.1, +np.inf, 30, 11, 0, 0, 95, 13.4])

#Set of column indices
col_indices = np.arange(N)
print(col_indices)
print(col_indices[13])

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13]
13


In [ ]:
#Declaring the decision variables in model
model.x = Var(col_indices)

In [ ]:
#Creating a constraint list to hold multiple constraints
model.constraints = ConstraintList()

#Adding model constraints one by one to the list
#first add the constraint p1x1 + p2x2 +.......+ pNxN == b1
model.constraints.add(sum(constr1_coef_p[j]*model.x[j] for j in col_indices) <= constr1_rhs_b1)

#second add the constraints q1x1 + qzx2 +......+ qNxN == b2
model.constraints.add(sum(constr2_coef_p[j]*model.x[j] for j in col_indices) == constr2_rhs_b2)

In [ ]:
#add the upper and lower bounds for the variables
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
#add the model objective
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense = minimize)

In [ ]:
#print the whole model
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   14 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}

1 Var Declarations
    x : Size=14, Index=x_index
        Key : Lower   : Value : Upper : Fixed : Stale : Domain
          0 :     2.0 :  None :  None : False :  True :  Reals
          1 :    -3.0 :  None :   6.0 : False :  True :  Reals
          2 :    None :  None :   2.5 : False :  True :  Reals
          3 :     1.0 :  None :   3.4 : False :  True :  Reals
          4 :     7.0 :  None :  None : False :  True :  Reals
          5 :     3.0 :  None :   6.0 : False :  True :  Reals
          6 :    -9.0 :  None :   4.1 : False :  True :  Reals
          7 :     0.0 :  None :  None : False :  True :  Reals
          8 :     1.0 :  No

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -14246.25704375
  Upper bound: -14246.25704375
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 15
  Number of nonzeros: 25
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.04184460639953613
# ----------------------------------------------------------
#   Solution Information
# ---------------

In [ ]:
print('\nObjective = ', model.objective())
print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)
print('\nConstraints')
model.constraints.display()


Objective =  -14246.257043750004

Decision Variables
x[ 0 ] =  2039.2619375
x[ 1 ] =  -3.0
x[ 2 ] =  -1623.2369375
x[ 3 ] =  1.0
x[ 4 ] =  7.0
x[ 5 ] =  3.0
x[ 6 ] =  -9.0
x[ 7 ] =  0.0
x[ 8 ] =  1.0
x[ 9 ] =  -25.245
x[ 10 ] =  0.0
x[ 11 ] =  -6.0
x[ 12 ] =  -30.5
x[ 13 ] =  -10.8

Constraints
constraints : Size=2
    Key : Lower : Body               : Upper
      1 :  None : 343.47999999999985 : 343.48
      2 :  66.2 :  66.19999999999925 :   66.2


**Report of exercise 1 part 2:**

*   **The values of the decision variables at optimal point are:**

   x1 = 2039.2619375

   x2 = -3.0

   x3 = -1623.2369375

   x4 =  1.0

   x5 = 7.0

   x6 = 3.0

   x7 = -9.0

   x8 = 0.0

   x9 = 1.0

   x10 = -25.245

   x11 = 0.0

   x12 = -6.0

   x13 = -30.5

   x14 = -10.8

*   **The value of the objective function at optimal point is:**

   -14246.257043750004




